File: ios/Classes/conference/peercontrol/PlanetKitFlutterPeerControlPlugin.swift
// Copyright 2024 LINE Plus Corporation
//
// LINE Plus Corporation licenses this file to you under the Apache License,
// version 2.0 (the "License"); you may not use this file except in compliance
// with the License. You may obtain a copy of the License at:
//
//   https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// License for the specific language governing permissions and limitations
// under the License.

import Foundation
import PlanetKit
import Flutter

extension PlanetKitPeerControl: PluginInstance {
    var instanceId: String {
        return "\(Unmanaged<AnyObject>.passUnretained(self).toOpaque())"
    }
}

class PlanetKitFlutterPeerControlPlugin {
    let nativeInstances: PlanetKitFlutterNativeInstances
    let eventStreamHandler: PlanetKitFlutterStreamHandler
    
    init(nativeInstances: PlanetKitFlutterNativeInstances, eventStreamHandler: PlanetKitFlutterStreamHandler) {
        self.nativeInstances = nativeInstances
        self.eventStreamHandler = eventStreamHandler
    }
    
    func register(call: FlutterMethodCall, result: @escaping FlutterResult) {
        PlanetKitLog.v("#flutter \(#function) \(String(describing: call.arguments))")
        let id = call.arguments as! String
        
        guard let peerControl = nativeInstances.get(key: id) as? PlanetKitPeerControl else {
            PlanetKitLog.e("#flutter \(#function) peer control not found \(id)")
            result(false)
            return
        }
        
        peerControl.register(self) { success in
            PlanetKitLog.v("#flutter \(#function) peeer control register result: \(success)")
            result(success)
        }
    }
    
    func unregister(call: FlutterMethodCall, result: @escaping FlutterResult) {
        PlanetKitLog.v("#flutter \(#function) \(String(describing: call.arguments))")
        let id = call.arguments as! String
        
        guard let peerControl = nativeInstances.get(key: id) as? PlanetKitPeerControl else {
            PlanetKitLog.e("#flutter \(#function) peer control not found \(id)")
            result(false)
            return
        }
        
        peerControl.unregister() { success in
            PlanetKitLog.v("#flutter \(#function) peeer control unregister result: \(success)")
            result(success)
        }
    }
    
    func startVideo(call: FlutterMethodCall, result: @escaping FlutterResult) {
        PlanetKitLog.v("#flutter \(#function) \(String(describing: call.arguments))")
        let param = PlanetKitFlutterPlugin.decodeMethodCallArg(call: call, codable: PeerControlParams.StartVideo.self)
        
        guard let peerControl = nativeInstances.get(key: param.id) as? PlanetKitPeerControl else {
            PlanetKitLog.e("#flutter \(#function) peer control not found \(param.id)")
            result(false)
            return
        }
        
        guard let view = PlanetKitFlutterVideoViews.shared.getView(id: param.viewId) else {
            PlanetKitLog.e("#flutter \(#function) view not found \(param.viewId)")
            result(false)
            return
        }
                
        peerControl.startVideo(maxResolution: param.maxResolution, delegate: view.delegate) { success in
            PlanetKitLog.v("#flutter \(#function) startVideo result \(success)")
            result(success)
        }
    }
    
    func stopVideo(call: FlutterMethodCall, result: @escaping FlutterResult) {
        PlanetKitLog.v("#flutter \(#function) \(String(describing: call.arguments))")
        let param = PlanetKitFlutterPlugin.decodeMethodCallArg(call: call, codable: PeerControlParams.StopVideo.self)
        
        guard let peerControl = nativeInstances.get(key: param.id) as? PlanetKitPeerControl else {
            PlanetKitLog.e("#flutter \(#function) peer control not found \(param.id)")
            result(false)
            return
        }
                
        peerControl.stopVideo() { success in
            PlanetKitLog.v("#flutter \(#function) stopVideo result \(success)")
            result(success)
        }
    }
    
    func startScreenShare(call: FlutterMethodCall, result: @escaping FlutterResult) {
        PlanetKitLog.v("#flutter \(#function) \(String(describing: call.arguments))")
        let param = PlanetKitFlutterPlugin.decodeMethodCallArg(call: call, codable: PeerControlParams.StartScreenShare.self)
        
        guard let peerControl = nativeInstances.get(key: param.id) as? PlanetKitPeerControl else {
            PlanetKitLog.e("#flutter \(#function) peer control not found \(param.id)")
            result(false)
            return
        }
        
        guard let view = PlanetKitFlutterVideoViews.shared.getView(id: param.viewId) else {
            PlanetKitLog.e("#flutter \(#function) view not found \(param.viewId)")
            result(false)
            return
        }
                
        peerControl.startScreenShare(delegate: view.delegate) { success in
            PlanetKitLog.v("#flutter \(#function) startVideo result \(success)")
            result(success)
        }
    }
    
    func stopScreenShare(call: FlutterMethodCall, result: @escaping FlutterResult) {
        PlanetKitLog.v("#flutter \(#function) \(String(describing: call.arguments))")
        let param = PlanetKitFlutterPlugin.decodeMethodCallArg(call: call, codable: PeerControlParams.StartScreenShare.self)
        
        guard let peerControl = nativeInstances.get(key: param.id) as? PlanetKitPeerControl else {
            PlanetKitLog.e("#flutter \(#function) peer control not found \(param.id)")
            result(false)
            return
        }
        
                
        peerControl.stopScreenShare() { success in
            PlanetKitLog.v("#flutter \(#function) stopScreenShare result \(success)")
            result(success)
        }
    }
}

extension PlanetKitFlutterPeerControlPlugin: PlanetKitPeerControlDelegate {
    func didMuteMic(_ peerControl: PlanetKitPeerControl) {
        DispatchQueue.main.async {
            PlanetKitLog.v("#flutter \(#function)")
            let event = PeerControlEvents.MicMuteEvent(id: peerControl.instanceId)
            let encodedEvent = PlanetKitFlutterPlugin.encode(data: event)
            self.eventStreamHandler.eventSink?(encodedEvent)
        }
    }
    
    func didUnmuteMic(_ peerControl: PlanetKitPeerControl) {
        DispatchQueue.main.async {
            PlanetKitLog.v("#flutter \(#function)")
            let event = PeerControlEvents.MicUnmuteEvent(id: peerControl.instanceId)
            let encodedEvent = PlanetKitFlutterPlugin.encode(data: event)
            self.eventStreamHandler.eventSink?(encodedEvent)
        }
    }
    
    func didHold(_ peerControl: PlanetKitPeerControl, reason: String) {
        DispatchQueue.main.async {
            PlanetKitLog.v("#flutter \(#function)")
            let event = PeerControlEvents.HoldEvent(id: peerControl.instanceId, reason: reason)
            let encodedEvent = PlanetKitFlutterPlugin.encode(data: event)
            self.eventStreamHandler.eventSink?(encodedEvent)
        }
    }
    
    func didUnhold(_ peerControl: PlanetKitPeerControl) {
        DispatchQueue.main.async {
            PlanetKitLog.v("#flutter \(#function)")
            let event = PeerControlEvents.UnholdEvent(id: peerControl.instanceId)
            let encodedEvent = PlanetKitFlutterPlugin.encode(data: event)
            self.eventStreamHandler.eventSink?(encodedEvent)
        }
    }
    
    func didDisconnect(_ peerControl: PlanetKitPeerControl) {
        DispatchQueue.main.async {
            PlanetKitLog.v("#flutter \(#function)")
            let event = PeerControlEvents.DisconnectEvent(id: peerControl.instanceId)
            let encodedEvent = PlanetKitFlutterPlugin.encode(data: event)
            self.eventStreamHandler.eventSink?(encodedEvent)
        }
    }
    
    func didUpdateAudioDescription(_ peerControl: PlanetKit.PlanetKitPeerControl, description: PlanetKit.PlanetKitPeerAudioDescription) {
        DispatchQueue.main.async {
            let event = PeerControlEvents.UpdateAudioDescriptionEvent(id: peerControl.instanceId, averageVolumeLevel: Int(description.averageVolumeLevel))
            let encodedEvent = PlanetKitFlutterPlugin.encode(data: event)
            self.eventStreamHandler.eventSink?(encodedEvent)
        }
    }
    
    func didUpdateVideo(_ peerControl: PlanetKit.PlanetKitPeerControl, subgroup: PlanetKit.PlanetKitSubgroup, status: PlanetKit.PlanetKitVideoStatus) {
        DispatchQueue.main.async {
            PlanetKitLog.v("#flutter \(#function)")
            let event = PeerControlEvents.UpdateVideoEvent(id: peerControl.instanceId, status: status)
            let encodedEvent = PlanetKitFlutterPlugin.encode(data: event)
            self.eventStreamHandler.eventSink?(encodedEvent)
        }
    }
    
    func didUpdateScreenShare(_ peerControl: PlanetKitPeerControl, subgroup: PlanetKitSubgroup, status: PlanetKitScreenShareStatus) {
        DispatchQueue.main.async {
            PlanetKitLog.v("#flutter \(#function)")
            let event = PeerControlEvents.UpdateScreenShareEvent(id: peerControl.instanceId, state: status.state)
            let encodedEvent = PlanetKitFlutterPlugin.encode(data: event)
            self.eventStreamHandler.eventSink?(encodedEvent)
        }
    }
}

File: android/src/main/kotlin/com/example/planet_kit_flutter/conference/peerControl/PlanetKitFlutterPeerControlPlugin.kt
// Copyright 2024 LINE Plus Corporation
//
// LINE Plus Corporation licenses this file to you under the Apache License,
// version 2.0 (the "License"); you may not use this file except in compliance
// with the License. You may obtain a copy of the License at:
//
//   https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// License for the specific language governing permissions and limitations
// under the License.

package com.example.planet_kit_flutter.conference.peerControl

import android.os.Handler
import android.os.Looper
import android.util.Log
import com.example.planet_kit_flutter.PlanetKitFlutterNativeInstances
import com.example.planet_kit_flutter.PlanetKitFlutterStreamHandler
import com.example.planet_kit_flutter.videoView.PlanetKitFlutterVideoViews
import com.google.gson.Gson
import com.linecorp.planetkit.PlanetKit
import com.linecorp.planetkit.audio.PlanetKitAudioDescription
import com.linecorp.planetkit.session.conference.PlanetKitPeerControl
import com.linecorp.planetkit.session.conference.subgroup.PlanetKitConferencePeer
import com.linecorp.planetkit.video.PlanetKitScreenShareState
import com.linecorp.planetkit.video.PlanetKitVideoStatus
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel


class PlanetKitFlutterPeerControlPlugin(
    private val eventStreamHandler: PlanetKitFlutterStreamHandler,
    private val nativeInstances: PlanetKitFlutterNativeInstances,
    private val gson: Gson
) {
    fun register(call: MethodCall, result: MethodChannel.Result) {
        Log.d("FlutterPlugin", "register ${call.method} ${call.arguments}")
        val id = call.arguments as String

        val peerControl = nativeInstances.get(id) as? PlanetKitPeerControl
        if (peerControl == null) {
            Log.e("FlutterPlugin", "register peer control not found $id")
            result.success(false)
            return
        }

        val ret = peerControl.register(object : PlanetKitPeerControl.PeerControlListener {
            override fun onMicMuted(peer: PlanetKitConferencePeer) {
                Handler(Looper.getMainLooper()).post {
                    Log.d("FlutterPlugin", "onMicMute")
                    val event = PeerControlEvents.MicMuteEvent(peerControl.hashCode().toString())
                    val json = gson.toJson(event)
                    eventStreamHandler.eventSink?.success(json)
                }
            }

            override fun onMicUnmuted(peer: PlanetKitConferencePeer) {
                Handler(Looper.getMainLooper()).post {
                    Log.d("FlutterPlugin", "onMicUnmuted")
                    val event = PeerControlEvents.MicUnmuteEvent(peerControl.hashCode().toString())
                    val json = gson.toJson(event)
                    eventStreamHandler.eventSink?.success(json)
                }
            }

            override fun onHold(peer: PlanetKitConferencePeer, reason: String?) {
                Handler(Looper.getMainLooper()).post {
                    Log.d("FlutterPlugin", "onHold")
                    val event =
                        PeerControlEvents.HoldEvent(peerControl.hashCode().toString(), reason)
                    val json = gson.toJson(event)
                    eventStreamHandler.eventSink?.success(json)
                }
            }

            override fun onUnhold(peer: PlanetKitConferencePeer) {
                Handler(Looper.getMainLooper()).post {
                    Log.d("FlutterPlugin", "onUnhold")
                    val event = PeerControlEvents.UnholdEvent(peerControl.hashCode().toString())
                    val json = gson.toJson(event)
                    eventStreamHandler.eventSink?.success(json)
                }
            }

            override fun onDisconnected(peer: PlanetKitConferencePeer) {
                Handler(Looper.getMainLooper()).post {
                    Log.d("FlutterPlugin", "onDisconnected")
                    val event = PeerControlEvents.DisconnectEvent(peerControl.hashCode().toString())
                    val json = gson.toJson(event)
                    eventStreamHandler.eventSink?.success(json)
                }
            }

            override fun onAudioDescriptionUpdated(
                peer: PlanetKitConferencePeer,
                audioDescription: PlanetKitAudioDescription
            ) {
                Handler(Looper.getMainLooper()).post {
                    val event = PeerControlEvents.UpdateAudioDescriptionEvent(
                        peerControl.hashCode().toString(), audioDescription.averageVolumeLevel
                    )
                    val json = gson.toJson(event)
                    eventStreamHandler.eventSink?.success(json)
                }
            }

            override fun onVideoUpdated(
                peer: PlanetKitConferencePeer,
                videoStatus: PlanetKitVideoStatus,
                subgroupName: String?
            ) {
                Handler(Looper.getMainLooper()).post {
                    Log.d("FlutterPlugin", "onVideoUpdated")
                    val event = PeerControlEvents.UpdateVideoEvent(
                        peerControl.hashCode().toString(), videoStatus
                    )
                    val json = gson.toJson(event)
                    eventStreamHandler.eventSink?.success(json)
                }
            }

            override fun onScreenShareUpdated(
                peer: PlanetKitConferencePeer,
                state: PlanetKitScreenShareState,
                subgroupName: String?
            ) {
                Handler(Looper.getMainLooper()).post {
                    Log.d("FlutterPlugin", "onScreenShareUpdated")
                    val event = PeerControlEvents.UpdateScreenShareEvent(
                        peerControl.hashCode().toString(), state
                    )
                    val json = gson.toJson(event)
                    eventStreamHandler.eventSink?.success(json)
                }
            }

        }) { res ->
            Log.d("FlutterPlugin", "register peer control result: ${res.isSuccessful}")
            result.success(res.isSuccessful)
        }

        if (!ret) {
            Log.d("FlutterPlugin", "register platform api  returned false")
            result.success(false)
        }
    }

    fun unregister(call: MethodCall, result: MethodChannel.Result) {
        Log.d("FlutterPlugin", "unregister ${call.method} ${call.arguments}")
        val id = call.arguments as String

        val peerControl = nativeInstances.get(id) as? PlanetKitPeerControl
        if (peerControl == null) {
            Log.e("FlutterPlugin", "unregister peer control not found $id")
            result.success(false)
            return
        }

        val ret = peerControl.unregister { res ->
            Log.d("FlutterPlugin", "unregister peer control result: ${res.isSuccessful}")
            result.success(res.isSuccessful)
        }

        if (!ret) {
            Log.d("FlutterPlugin", "unregister platform api  returned false")
            result.success(false)
        }
    }

    fun startVideo(call: MethodCall, result: MethodChannel.Result) {
        Log.d("FlutterPlugin", "startVideo ${call.method} ${call.arguments}")
        val param =
            gson.fromJson(call.arguments.toString(), PeerControlParams.StartVideoParam::class.java)

        val peerControl = nativeInstances.get(param.id) as? PlanetKitPeerControl
        if (peerControl == null) {
            Log.e("FlutterPlugin", "startVideo peer control not found ${param.id}")
            result.success(false)
            return
        }

        val ret = peerControl.startVideo(param.maxResolution)

        if (!ret) {
            Log.d("FlutterPlugin", "startVideo platform api  returned false")
            result.success(false)
            return
        }

        val videoView = PlanetKitFlutterVideoViews.getView(param.viewId)

        if (videoView == null) {
            Log.d("FlutterPlugin", "failed to find the view for ${param.viewId}")
            result.success(false)
            return
        }

        val conference = PlanetKit.getConference()

        if (conference == null) {
            Log.d("FlutterPlugin", "conference is null")
            result.success(false)
            return
        }

        conference.addPeerVideoView(peerControl.peer.user, videoView.videoView)
        PlanetKitFlutterVideoViews.retain(param.viewId)
        result.success(true)
    }

    fun stopVideo(call: MethodCall, result: MethodChannel.Result) {
        Log.d("FlutterPlugin", "stopVideo ${call.method} ${call.arguments}")
        val param =
            gson.fromJson(call.arguments.toString(), PeerControlParams.StopVideoParam::class.java)

        val peerControl = nativeInstances.get(param.id) as? PlanetKitPeerControl
        if (peerControl == null) {
            Log.e("FlutterPlugin", "stopVideo peer control not found ${param.id}")
            result.success(false)
            return
        }

        val ret = peerControl.stopVideo()

        if (!ret) {
            Log.d("FlutterPlugin", "stopVideo platform api  returned false")
            result.success(false)
            PlanetKitFlutterVideoViews.release(param.viewId)
            return
        }

        val videoView = PlanetKitFlutterVideoViews.getView(param.viewId)

        if (videoView == null) {
            Log.d("FlutterPlugin", "failed to find the view for ${param.viewId}")
            result.success(false)
            return
        }

        val conference = PlanetKit.getConference()

        if (conference == null) {
            Log.d("FlutterPlugin", "conference is null")
            result.success(false)
            return
        }

        conference.removePeerVideoView(peerControl.peer.user, videoView.videoView)
        PlanetKitFlutterVideoViews.release(param.viewId)
        result.success(true)
    }

    fun startScreenShare(call: MethodCall, result: MethodChannel.Result) {
        Log.d("FlutterPlugin", "startScreenShare ${call.method} ${call.arguments}")
        val param =
            gson.fromJson(call.arguments.toString(), PeerControlParams.StartScreenShareParam::class.java)

        val peerControl = nativeInstances.get(param.id) as? PlanetKitPeerControl
        if (peerControl == null) {
            Log.e("FlutterPlugin", "startScreenShare peer control not found ${param.id}")
            result.success(false)
            return
        }

        val ret = peerControl.startScreenShare()

        if (!ret) {
            Log.d("FlutterPlugin", "startScreenShare platform api  returned false")
            result.success(false)
            return
        }

        val videoView = PlanetKitFlutterVideoViews.getView(param.viewId)

        if (videoView == null) {
            Log.d("FlutterPlugin", "failed to find the view for ${param.viewId}")
            result.success(false)
            return
        }

        val conference = PlanetKit.getConference()

        if (conference == null) {
            Log.d("FlutterPlugin", "conference is null")
            result.success(false)
            return
        }

        conference.addPeerScreenShareView(peerControl.peer.user, videoView.videoView)
        PlanetKitFlutterVideoViews.retain(param.viewId)
        result.success(true)
    }

    fun stopScreenShare(call: MethodCall, result: MethodChannel.Result) {
        Log.d("FlutterPlugin", "stopScreenShare ${call.method} ${call.arguments}")
        val param =
            gson.fromJson(call.arguments.toString(), PeerControlParams.StopScreenShareParam::class.java)

        val peerControl = nativeInstances.get(param.id) as? PlanetKitPeerControl
        if (peerControl == null) {
            Log.e("FlutterPlugin", "stopScreenShare peer control not found ${param.id}")
            result.success(false)
            return
        }

        val ret = peerControl.stopScreenShare()

        if (!ret) {
            Log.d("FlutterPlugin", "stopScreenShare platform api  returned false")
            result.success(false)
            return
        }

        val videoView = PlanetKitFlutterVideoViews.getView(param.viewId)

        if (videoView == null) {
            Log.d("FlutterPlugin", "failed to find the view for ${param.viewId}")
            result.success(false)
            return
        }

        val conference = PlanetKit.getConference()

        if (conference == null) {
            Log.d("FlutterPlugin", "conference is null")
            result.success(false)
            return
        }

        conference.removePeerScreenShareView(peerControl.peer.user, videoView.videoView)
        PlanetKitFlutterVideoViews.release(param.viewId)
        result.success(true)
    }
}


